### ISSUE TITLE
indent rule enhancements

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#581]
Similar to [ESLint](http://eslint.org/docs/rules/indent), we should be able to enforce 2-space or 4-space indentation via options like so:
- [ ] `[true, "spaces", 2]`
- [ ] `[true, "spaces", 4]`

In addition, it'd be nice to have a `"detect"` option (as suggested in #122) which tries to guess the indentation style in a file and enforce consistency:
- [ ] `[true, "detect"]`

**Comment by IlanFrumer** [comment]
+1

**Comment by nikklassen** [comment]
+1 for enforcing number of spaces.

**Comment by allenhwkim** [comment]
+1

**Comment by glen-84** [comment]
Would this only work if the indentation is divisible by the number of spaces configured?

```js
{
    theObject.something()
             .more(); // 13 spaces.
}
```

The above (align the periods) method wouldn't be supported?

**Comment by JKillian** [comment]
@glen-84 I see two possibilities: 

* A line by line check of indentation (where your example wouldn't be allowed
* A statement by statement check of indentation. Since in your example, `theObject.something().more()` is all one statement, it would only check the indentation level at the beginning of it.

I think they both have pros and cons, but the second one is probably preferred here.

**Comment by nwhite89** [comment]
:+1:

**Comment by jppellerin** [comment]
+1

**Comment by adidahiya** [comment]
+1 option 2, checking statement-by-statement

**Comment by glen-84** [comment]
@JKillian Thanks. The second option looks good. =)

**Comment by tengri** [comment]
+1 option 1

**Comment by leocaseiro** [comment]
A good way is the ESLINT one http://eslint.org/docs/rules/indent

**Comment by radarsu** [comment]
+9000 SSJ3
Please, guys, implement 4-spaces indent option ASAP. I started programming in typescript a lot and 2-spaces default indent feels so bad for me. For some people it's really hard to see indent level of some code with only 2 spaces...

**Comment by leocaseiro** [comment]
@radarsu you might want to try use [.editorconfig](http://editorconfig.org/) for the time being.

**Comment by JKillian** [comment]
@radarsu Right now the indentation rule just figures out if you have an indentation error based upon how your code aligns with the previous statement. You're able to use four space indentation and make the rule happy (in fact, that's what the TSLint source code does)

**Comment by patsissons** [comment]
@JKillian any idea how hard this would be to tackle? i would love to start enforcing the indent size. I'm open to hacking away at it if it's not too much of a time sink (only have so much time over the next week).

**Comment by JKillian** [comment]
Hard to say how difficult it would be - I expect it wouldn't be simple at the least haha. Definitely go for it though!

**Comment by mcriss** [comment]
+1

**Comment by strike-sbolel** [comment]
+1

**Comment by glen-84** [comment]
@mcriss @strike-sbolel Please stop adding `+1` comments.

**Comment by noxan** [comment]
Looks like similar efforts are being made here https://github.com/buzinas/tslint-eslint-rules/issues/74

**Comment by glen-84** [comment]
I've [changed my mind](https://github.com/palantir/tslint/issues/581#issuecomment-186654375), as option 2 would allow code like:

```ts
{
    theObject.something()
             .more() // 13 spaces.
           .wrong()
                    .moreWrong();
}
```

In an ideal world, we'd use option 1 with a setting to allow (or enforce) "wrapped" lines to be positioned below the last period in the previous line.

```ts
{
    theObject.something()
             .more()  // 13 spaces.
             .right() // Must be the same as above now, because the previous line used this indentation mode.
             .moreRight();
}
```

Or:

```ts
{
    theObject.something()
        .more() // (4 x 2) spaces.
        .right()
        .moreRight();
}
```

**Comment by jmlopez-rod** [comment]
Merry Christmas y'all!!!

https://github.com/buzinas/tslint-eslint-rules/blob/master/src/docs/rules/terIndentRule.md

**Comment by owais** [comment]
Can we release an update please?

**Comment by adidahiya** [comment]
@owais the rule hasn't been implemented in core TSLint yet. That link is to an external repo with custom rules.

**Comment by owais** [comment]
Oh. I didn't realize I was looking at another repo. :) Thanks for the link. I think I might use that.

**Comment by pudgeball** [comment]
@adidahiya Is this still something that is outstanding, and is it worth me opening a PR? Or would it be better to use the custom rules you've linked to?

**Comment by adidahiya** [comment]
@pudgeball definitely would be happy to accept a PR (as indicated by the label)

**Comment by nchen63** [comment]
note that #2723 doesn't fix this issue. It only fixes tabs -> spaces or spaces -> tabs, but doesn't fix x spaces -> y spaces

**Comment by dimitriy-k** [comment]
still no updates on indent size?... request is from 2015

**Comment by adidahiya** [comment]
@dimitriy-k #2723 just merged, it'll be in the next release.

**Comment by dolanmiu** [comment]
What will this rule be called? 😄

**Comment by adidahiya** [comment]
@dolanmiu it's not a new rule. It's implemented exactly as noted in the original issue description. Just click through to the PR I linked.

**Comment by dolanmiu** [comment]
@adidahiya I am using the current latest version of tslint `5.3.2`

And using the following:

```json
    "indent": [
      true,
      "spaces",
      4
    ],
```

I have a file which is 2 all 2 spaces, but it did not detect it! Am I doing something wrong, or is there a bug?

**Comment by adidahiya** [comment]
@dolanmiu yeah that looks like a new issue, I also noticed #2723 did not work like I expected. I filed https://github.com/palantir/tslint/issues/2814

**Comment by liul85** [comment]
any updates? 

`"indent": [true, "spaces", 4]` not working for me

**Comment by leocaseiro** [comment]
@liul85 try install [tslint-eslint-rules](https://github.com/buzinas/tslint-eslint-rules) and use the [terIndentRule](https://github.com/buzinas/tslint-eslint-rules/blob/master/src/docs/rules/terIndentRule.md)

**Comment by liul85** [comment]
@leocaseiro thanks for sharing, I will give a try.

**Comment by anton-aurea** [comment]
Still not working in tslint 5.11.0

**Comment by rHilkner** [comment]
`"indent": [true, "spaces", 4]` still not working...
